jenkinsapi/plugins.py

```diff
@@ -57,6 +83,173 @@ def __getitem__(self, plugin_name):
             raise UnknownPlugin(plugin_name)
 
+    def __setitem__(self, shortName, plugin):
+        """
+        Installs plugin in Jenkins.
+
+        If plugin already exists - this method is going to uninstall the existing
+        plugin and install the specified version if it is not already installed.
+
+        :param shortName: Plugin ID
+        :param plugin a Plugin object to be installed.
+        """
+        if self.plugin_version_already_installed(plugin):
+            return
+        if plugin.is_latest(self.update_center_dict):
+            self._install_plugin_from_updatecenter(plugin)
+        else:
+            self._install_specific_version(plugin)
+        self._wait_until_plugin_installed(plugin)
+
+    def _install_plugin_from_updatecenter(self, plugin):
+        """
+        Latest versions of plugins can be installed from the update center (and don't need a restart.)
+        """
+        xml_str = plugin.get_attributes()
+        url = (
+            '%s/pluginManager/installNecessaryPlugins' % self.jenkins_obj.baseurl
+        )
+        self.jenkins_obj.requester.post_xml_and_confirm_status(url, data=xml_str)
+
+    @property
+    def update_center_install_status(self):
+        """
+        Jenkins 2.x specific
+        """
+        url = "%s/updateCenter/installStatus" % self.jenkins_obj.baseurl
+        status = self.jenkins_obj.requester.get_url(url)
+        if status.status_code == 404:
+            raise JenkinsAPIException('update_center_install_status not available for Jenkins 1.X')
+        return status.json()
+
+    @property
+    def restart_required(self):
+        """
+        Call after plugin installation to check if Jenkins requires a restart
+        """
+        try:
+            jobs = self.update_center_install_status['data']['jobs']
+        except JenkinsAPIException:
+            return True  # Jenkins 1.X has no update_center
+        return any([job for job in jobs if job['requiresRestart'] == 'true'])
+
+    def _install_specific_version(self, plugin):
+        """
+        Plugins that are not the latest version have to be uploaded.
+        """
+        download_link = plugin.get_download_link(update_center_dict=self.update_center_dict)
+        downloaded_plugin = self._download_plugin(download_link)
+        plugin_dependencies = self._get_plugin_dependencies(downloaded_plugin)
+        log.debug("Installing dependencies for plugin '%s'", plugin.shortName)
+        self.jenkins_obj.install_plugins(plugin_dependencies)
+        url = ('%s/pluginManager/uploadPlugin' % self.jenkins_obj.baseurl)
+        requester = self.jenkins_obj.requester
+        downloaded_plugin.seek(0)
+        requester.post_and_confirm_status(url, files={'file': ('plugin.hpi', downloaded_plugin)}, data={}, params={})
+
+    def _get_plugin_dependencies(self, downloaded_plugin):
+        """
+        Returns a list of all dependencies for a downloaded plugin
+        """
+        plugin_dependencies = []
```

Comment:
That would be the dependency for the latest version only, right?
I'm OK for doing this through the API, but keep in mind that this code is only called when requesting a plugin that is not the latest version.